arduino-ide-extension/src/node/board-discovery.ts

```diff
@@ -334,21 +242,130 @@ export class BoardDiscovery
+      const key = Port.keyOf(port);
+
+      if (eventType === EventType.Add) {
+        if (newState[key]) {
+          const [, knownBoards] = newState[key];
+          this.logger.warn(
+            `Port '${Port.toString(
+              port
+            )}' was already available. Known boards before override: ${JSON.stringify(
+              knownBoards
+            )}`
+          );
+        }
+        newState[key] = [port, boards];
+      } else if (eventType === EventType.Remove) {
+        if (!newState[key]) {
+          this.logger.warn(
+            `Port '${Port.toString(port)}' was not available. Skipping`
+          );
+          return;
+        }
+        delete newState[key];
+      }
+
+      const event: AttachedBoardsChangeEvent = {
+        oldState: {
+          ...AvailablePorts.split(oldState),
+        },
+        newState: {
+          ...AvailablePorts.split(newState),
+        },
+      };
+
+      this._availablePorts = newState;
+      this.notificationService.notifyAttachedBoardsDidChange(event);
     }
-    return attachedBoards;
   }
 
-  getAvailablePorts(state: AvailablePorts = this.availablePorts): Port[] {
-    const availablePorts: Port[] = [];
-    for (const portID of Object.keys(state)) {
-      const [port] = state[portID];
-      availablePorts.push(port);
+  private fromRpc(detectedPort: RpcDetectedPort): DetectedPort {
+    const rpcPort = detectedPort.getPort();
+    const port = rpcPort && this.fromRpcPort(rpcPort);
+    const boards = detectedPort.getMatchingBoardsList().map(
+      (board) =>
+        ({
+          fqbn: board.getFqbn(),
+          name: board.getName() || Unknown,
+          port,
+        } as Board)
+    );
+    return {
+      boards,
+      port,
+    };
+  }
+
+  private fromRpcPort(rpcPort: RpcPort): Port {
+    const port = {
+      address: rpcPort.getAddress(),
+      addressLabel: rpcPort.getLabel(),
+      protocol: rpcPort.getProtocol(),
+      protocolLabel: rpcPort.getProtocolLabel(),
+      properties: Port.Properties.create(rpcPort.getPropertiesMap().toObject()),
+    };
+    return port;
+  }
+}
+
+enum EventType {
+  Add,
+  Remove,
+  Quit,
+}
+namespace EventType {
+  export function parse(type: string): EventType {
+    const normalizedType = type.toLowerCase();
+    switch (normalizedType) {
+      case 'add':
+        return EventType.Add;
+      case 'remove':
+        return EventType.Remove;
+      case 'quit':
+        return EventType.Quit;
+      default:
+        throw new Error(
+          `Unexpected 'BoardListWatchResponse' event type: '${type}.'`
+        );
     }
-    return availablePorts;
+  }
+}
+
+interface DetectedPort {
+  port: Port | undefined;
+  boards: Board[];
+}
+namespace DetectedPort {
+  export function isUnrecognizedNetwork(
+    arg: DetectedPort
+  ): arg is DetectedPort & { port: Port & { protocol: 'network' } } {
+    return !!arg.port && arg.port.protocol === 'network' && !arg.boards.length;
```

Comment:
I wasn't able to understand the purpose of this newly added protocol-specific code. Would you mind providing a quick summary of why it was added @kittaakos?

Reply:
It's a hack and will be removed. I have not yet pushed the changes to the remote.